### PR TITLE
CSharpForMarkup Part 2

### LINKED ISSUE #9905: [Bug] BindableLayout EmptyView doesn't work with a DataTemplateSelector
### Description

Attempting to use a `DataTemplateSelector` to return a `DataTemplate` to the `BindableLayout.EmptyViewTemplate` property results in a `TargetInvocationException` on iOS and Android.

It might be an unlikely scenario, but it'd be good if it was at least handled gracefully.

### Steps to Reproduce

1. Run the attached sample.
2. Browse to the "EmptyViewTemplate demo - Data template selector" page.

### Expected Behavior

DataTemplateSelector selects a DataTemplate, which defines the displayed empty view.

### Actual Behavior

`TargetInvocationException`.

### Basic Information

- Version with issue: 4.5-pre1
- IDE: VSMac 8.4.8
- Platform Target Frameworks:
  - iOS:  13.3
  - Android: API 28

### Reproduction Link

[BindableLayouts.zip](https://github.com/xamarin/Xamarin.Forms/files/4307283/BindableLayouts.zip)

### PR DESCRIPTION
**Pull request by VincentH-Net**
>## Note: C# Markup has been moved to the Xamarin Community Toolkit
>The features in this PR will not be included in Xamarin Forms. Instead they have been included in the [Xamarin Community Toolkit C# Markup](https://github.com/MicrosoftDocs/xamarin-communitytoolkit/blob/68fe0de991a99b364ed61e19bc280e3da94ff02e/docs/markup.md) release. Go check it out!

Adds support to C# Markup for:

- `MultiBinding`
- `MultiConverter`
- `BindableLayout`
- `RelativeLayout`
- `DynamicResource`

Improvements:
- `Bind` helpers are now more strongly typed, and `converterParameter` was removed from `Bind` overloads that did not use it
- `Font` helper now has `family` as first parameter (reduces the need to use named parameters across all `FontElement` extensions)
- `Padding` and `Paddings` helpers now work on all `IPaddingElement` view types
- New `Stars(double)` overload for defining rows and columns in Grids

### Issues Resolved ### 
Fix for issue encountered by new C# Markup unit tests:
`BindableLayout.SetEmptyViewTemplate` sets `EmptyView` instead of `EmptyViewTemplate`
Possibly this fixes #9905

### Multi-Binding Helpers
New overloads of the `Bind` helper offer support for  [multi-binding](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/app-fundamentals/data-binding/multibinding).

There are overloads that support 2, 3 or 4 bindings with a type-safe inline convertor. E.g:

```CSharp
new Label { }
    .Bind (Label.TextProperty,
        new Binding (nameof(vm.Name)),
        new Binding (nameof(vm.Score)),
        ((string name, bool score) v) => $"{v.name} Score: { v.score }"
    )
```
The value for all bindings are passed in as a `ValueTuple` with type-safe members.

You can also pass in a type-safe converter parameter:
```CSharp
new Label { }
    .Bind (Label.TextProperty,
        new Binding (nameof(vm.Name)),
        new Binding (nameof(vm.Score)),
        ((string name, int Score) v, bool winner) => $"{v.name} Score: { v.Score } Winner: { winner }",
        converterParameter: true
    )
```
Here `bool winner` gets the value from the `converterParameter`.

You can specify two-way conversion inline:
```CSharp
new Entry { }
    .Bind(Entry.TextProperty,
        new Binding (nameof(vm.Emoticon)),
        new Binding (nameof(vm.Repeat)),
        ((char emoticon, int repeat) v) => new string(v.emoticon, v.repeat),
        (string emoticons) => (emoticons[0], emoticons.Length)
    );
```
In the `convertBack` function you return the same `ValueTuple` that you receive in the `convert` function.
 
You can specify more than 4 bindings by passing in a multi-value converter, e.g. a `FuncMultiConverter`:
```CSharp
new Label { }
    .Bind(Label.TextProperty,
        new List<BindingBase> {
            new Binding(nameof(vm.Name)),
            new Binding(nameof(vm.Score))
        },
        new FuncMultiConverter<string, bool>(
            (object[] values, bool winner) => $"{values[0]} Score: { values[1] } Winner: { winner }"
        )
    )
```
This is not type-safe; you will need to cast the values to their type in the `convert` function.

The `FuncMultiConverter` classes implement `IMultiValueConverter`. The class used for any number of bindings is `FuncMultiConverter<TDest, TParam>`, which only specifies the destination type and the parameter type of the convertor. The binding values are passed as an `object[]`.

There are also type-safe generic overloads for `FuncMultiConverter` that take 2, 3 or 4 values (and optionally a convertor parameter). These classes pass the binding values in a type-safe `ValueTuple`.

### Bindable Layout Helpers
The `EmptyView`,  `EmptyViewTemplate`, `ItemsSource`, `ItemTemplate` and `ItemTemplateSelector` helpers offer support for [bindable layouts](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/layouts/bindable-layouts) on all `Layout<View>` types. E.g.:

```CSharp
new StackLayout { } 
    .ItemTemplate (() =>
        new Label { }
            .Bind (nameof(Item.Name))
        )
    .ItemsSource (vm.Items)
```

### RelativeLayout Helpers
The `Children` helper lets you to add constrained child views to a `RelativeLayout`.

To create constrained views from normal views, four helpers have been added: `Unconstrained`, `Constraints` and two `Constrain` overloads. Each overload returns a corresponding `*ConstrainedView` class, which offers a fluent API for setting constraints on `RelativeLayout` child views.

Constraints on a child view can be set with:
- A single Bounds expression
- Separate expressions for X, Y, Width and Height
- Separate `Constraint` instances for X, Y, Width and Height. Each of these Constraint instances has overloads for:
  - Constant
  - Relative to parent
  - Relative to view

Some examples:
```CSharp
new RelativeLayout { } .Children (
    new Label { } // Bounds constrained
        .Assign (out Label child0)
        .Constrain(() => new Rectangle(30, 20, layout.Height / 2, layout.Height / 4)),

    new Label { } // Expressions constrained
        .Constrain() .X      (() => 30)
                     .Y      (() => 20)
                     .Width  (() => layout.Height / 2)
                     .Height (() => layout.Height / 4),

    new Label { } // Constraints constrained - parent relative
        .Constraints() .X      (30)
                       .Y      (20)
                       .Width  (parent => parent.Height / 5)
                       .Height (parent => parent.Height / 10),

    new Label { } // Constraints constrained - view relative
        .Constraints() .X      (child0, (layout, view) => view.Bounds.Right + 10)
                       .Y      (child0, (layout, view) => view.Y)
                       .Width  (child0, (layout, view) => view.Width)
                       .Height (child0, (layout, view) => view.Height),
) .Assign (out layout)
```

### Dynamic Resource Helpers
The `DynamicResource`, `DynamicResources` and `RemoveDynamicResources` helpers add support for setting dynamic resources on an `Element`.

E.g.:
```CSharp
new Label { }
    .DynamicResource (Label.TextProperty, "TextKey")

new Label { }
    .DynamicResources((Label.TextProperty     , "TextKey"),
                      (Label.TextColorProperty, "ColorKey"));
```
### Platforms Affected ### 
- Core (all platforms)

### Behavioral/Visual Changes ###
None

### Before/After Screenshots ### 
Not applicable

### Testing Procedure ###
Fully automated; all added and changed code has **100% code coverage** from the included unit tests.

The new functionality was propotyped for usability in:
- MAUI [[Spec] C# Markup for MVU and MVVM](https://github.com/dotnet/maui/issues/119)
- [GitTrends](https://github.com/VincentH-Net/GitTrends/blob/master/GitTrends/Pages/ReferringSitesPage.cs) 
- [C# Markup versus Mobile Blazor Bindings in .NET MAUI](https://vincenth.net/blog/2020/07/11/c-markup-versus-mobile-blazor-bindings-in-net-maui)

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Comment by samhouts** [comment]
Hi @VincentH-Net! 

Thank you for continuing to improve this awesome feature! We want to keep the momentum going on building UI with C# code, and this is amazing work. 

We've been discussing this internally, and we think that this should move to the Xamarin Community Toolkit. We'd love to see continued innovation in this area, and the toolkit is the perfect place for that during this transition period between Xamarin.Forms and .NET MAUI.

Please join us on [Discord](https://aka.ms/dotnet-discord), if you haven't already, and we can discuss this further if you like!

Thanks again!

File: Xamarin.Forms.Core/BindableLayout.cs [hunk 1]
```diff
@@ -77,7 +77,7 @@ public static DataTemplate GetEmptyViewTemplate(BindableObject b)
 
 		public static void SetEmptyViewTemplate(BindableObject b, DataTemplate value)
 		{
-			b.SetValue(EmptyViewProperty, value);
+			b.SetValue(EmptyViewTemplateProperty, value);
```
Comment (VincentH-Net) [inline Xamarin.Forms.Core/BindableLayout.cs:80]: Fix for `BindableLayout.SetEmptyViewTemplate` sets `EmptyView` instead of `EmptyViewTemplate`

**Review by jsuarezruiz — CHANGES_REQUESTED** [review]
@VincentH-Net  We have C# Markup extensions in the Xamarin Community Toolkit now. Some of this PR chcanges is pending in the Toolkit or can we close it?

**Comment by VincentH-Net** [comment]
@jsuarezruiz  it can be closed, see the note in the pr description above:

> The features in this PR will not be included in Xamarin Forms. Instead they have been included in the Xamarin Community Toolkit C# Markup release. Go check it out!

**Pull request closed without merging**